src/main/java/org/truffleruby/extra/ConcurrentHashMapNodes.java

```diff
@@ -0,0 +1,300 @@
+/*
+ * Copyright (c) 2015, 2020 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 2.0, or
+ * GNU General Public License version 2, or
+ * GNU Lesser General Public License version 2.1.
+ */
+package org.truffleruby.extra;
+
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.object.Shape;
+import org.truffleruby.builtins.CoreMethod;
+import org.truffleruby.builtins.CoreMethodArrayArgumentsNode;
+import org.truffleruby.builtins.CoreModule;
+import org.truffleruby.builtins.YieldingCoreMethodNode;
+import org.truffleruby.core.basicobject.RubyBasicObject;
+import org.truffleruby.core.hash.RubyHash;
+import org.truffleruby.core.klass.RubyClass;
+import org.truffleruby.core.proc.RubyProc;
+import org.truffleruby.language.Nil;
+import org.truffleruby.language.NotProvided;
+import org.truffleruby.language.Visibility;
+import org.truffleruby.language.objects.AllocationTracing;
+
+import java.util.Map;
+
+@CoreModule(value = "TruffleRuby::ConcurrentHashMap", isClass = true)
+public class ConcurrentHashMapNodes {
+
+    @CoreMethod(names = { "__allocate__", "__layout_allocate__" }, constructor = true, visibility = Visibility.PRIVATE)
+    public abstract static class AllocateNode extends CoreMethodArrayArgumentsNode {
+
+        @Specialization
+        protected RubyConcurrentHashMap allocate(RubyClass rubyClass) {
+            final Shape shape = getLanguage().concurrentHashMapShape;
+            final RubyConcurrentHashMap instance = new RubyConcurrentHashMap(rubyClass, shape, null);
+            AllocationTracing.trace(instance, this);
+            return instance;
+        }
+    }
+
+    @CoreMethod(names = "initialize", optional = 1)
+    public abstract static class InitializeNode extends CoreMethodArrayArgumentsNode {
+
+        @Specialization
+        protected RubyConcurrentHashMap initializeNoValue(RubyConcurrentHashMap self, NotProvided options) {
+            return self;
+        }
+
+        @Specialization
+        protected RubyConcurrentHashMap initializeNil(RubyConcurrentHashMap self, Nil options) {
+            return self;
+        }
+
+        @Specialization
+        protected RubyConcurrentHashMap initializeWithOptions(RubyConcurrentHashMap self, RubyHash options) {
+            return self;
+        }
+
+        @Specialization(guards = { "!isNil(options)", "wasProvided(options)" })
+        protected RubyConcurrentHashMap initializeWithOptions(RubyConcurrentHashMap self, RubyBasicObject options) {
+            return self;
+        }
+    }
```

Comment:
We need to get rid of this code to handle options next - this whole method should go away and we should use the default initialiser.

Reply:
Since concurrent-ruby has `:initial_capacity` and `:load_factor` options I think we should at least accept them, and if possible pass them to the ConcurrentHashMap. It might be helpful to tune performance more finely.
So we could define `initialize` in Ruby code like
```ruby
def initialize(initial_capacity: nil, load_factor: nil, **options)
  Primitive.concurrent_hash_map_init(initial_capacity, load_factor)
end
```
in a `concurrent_hash_map.rb` file.

Reply:
@eregon It seems like both [`:initial_capacity`](https://github.com/wildmaples/concurrent-ruby/blob/3a0bbef9f51ea8e72886fe058dedc790567f8eeb/lib/concurrent/thread_safe/atomic_reference_cache_backend.rb#L370) and [`:load_factor`](https://github.com/wildmaples/concurrent-ruby/blob/3a0bbef9f51ea8e72886fe058dedc790567f8eeb/lib/concurrent/cache.rb#L172) is passed in within the options hash in `concurrent-ruby`. Unless I'm misunderstanding this completely, would this `initialize` work better? 

```
def initialize(options = nil)
  Primitive.concurrent_map_init(options[:initial_capacity], options[:load_factor])
end
```

Reply:
The options in `Concurrent::Map` are a bit of a mess - do they predate keyword arguments? They're also designed so you can ignore options you don't except. We've got a solution but it's not as simple as we'd like.

Reply:
Probably the code @wildmaples posted above + a `if options` might work.
We'll need to handle options not being passed, we could give sentionel/invalid values to detect them easily like say `0`:
```ruby
def initialize(options = nil)
  initial_capacity = options ? options[:initial_capacity] : 0
  load_factor = options ? options[:load_factor] : 0
  Primitive.concurrent_map_init(initial_capacity, load_factor)
end

```

Reply:
Ah, I misunderstood @wildmaples' question, yeah we can't use keyword arguments here because the caller passes `nil` if no options are passed.